### PR TITLE
Change ToHTML to also do one br per n

### PR DESCRIPTION
**Pull request by snan**

### CODE REVIEW

**Comment by dimkr** [comment]
Interesting, I prefer to wait with this until I hear complaints about weird formatting of outgoing posts. From what I see, Mastodon posts (by far the most common kind of posts) only use `p` tags and if I have to mimic the behavior of some other server, I choose Mastodon.

File: front/text/plain/convert.go [hunk 1]
```diff
@@ -132,11 +132,12 @@ func ToHTML(text string) string {
 		return text
 	}
 
+	text = strings.ReplaceAll(text, "\n\n", "</p><p>")
+	text = strings.ReplaceAll(text, "\n", "<br/>")
+
 	var b strings.Builder
-	for _, line := range lines {
```
Comment (dimkr) [inline front/text/plain/convert.go:136]: How about this instead?

```
	for _, line := range lines {
		if line == "" {
			w.WriteString("<br/>")
			continue
                }
 		b.WriteString("<p>")
		b.WriteString(line)
		b.WriteString("</p>")
	}
```

Each non-empty line is wrapped with <p></p>, otherwise replaced with <br/>. I took what seems to be a representative sample of posts from various servers, and this should mimic what the vast majority do. Don't forget that the input that passes through this function is plain text, not gemtext.

Comment (snan) [reply front/text/plain/convert.go:136]: Ah, no! Here's why:

So before, this input

    Lorem ipsum dolor
    sit amet

    consectetur adipisicing elit,
    sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua

would turn into

    Lorem ipsum dolor

    sit amet

    consectetur adipisicing elit,

    sed do eiusmod tempor incididunt ut

    labore et dolore magna aliqua

(And that's *after* the whitespace collapsing that Mastodon does. Without whitespace collapsing, there are even more blank lines in there.)

With my patch, the input displays the same on Mastodon as it does on the tootik side:


    Lorem ipsum dolor
    sit amet

    consectetur adipisicing elit,
    sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua

With your change, it would instead look like this

	Lorem ipsum dolor

	sit amet


	consectetur adipisicing elit,

	sed do eiusmod tempor incididunt ut

	labore et dolore magna aliqua


The explicit BR overrides the whitespace collapse making it even more weird-looking.

An argument could be made that it should become


    Lorem ipsum dolor sit amet

    consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

Since people are using this who are used to the way Gemini line wrapping works, I would argue against that, but a stronger case could be made for that than for inserting extra blank lines. Which Mastodon server don't do either.

Comment (dimkr) [reply front/text/plain/convert.go:136]: I see. This is more consistent with Mastodon at the cost of slightly weird posts if you're used gemtext (where most lines are actually paragraphs).

Many posts in Mastodon look like this:

```
line 1
line 2

line 3
```

But this is uncommon in gemtext, which is more likely to look like this:

```
line 1

line 2

line 3
```

If I have to choose between the two, it must be the first option. I'd add more tests, though, for weird cases like posts that begin with empty lines, or have nothing but empty lines.

Comment (snan) [reply front/text/plain/convert.go:136]: What I typically see on Mastodon and Fedi is the same as I see in gemtext:


    This is one paragraph.

    This is a second paragraph.

    Roses are red,
    violets are blue,
    the third paragraph's a poem
    through and through.

    This is the fourth paragraph.


As long as there was no brs (no poems), like this:

    This is one paragraph

    This is another paragraph

    This is a third, non-poetic paragraph,

    This is the fourth paragraph


The old FromHTML was preserving that on the gemside exactly:

    This is one paragraph

    This is another paragraph

    This is a third, non-poetic paragraph,

    This is the fourth paragraph


But with BRs, it would treat them as paragraphs, so the old version of FromHTML would do this to the poem version:


    This is one paragraph.

    This is a second paragraph.

    Roses are red,

    violets are blue,

    the third paragraph's a poem

    through and through.

    This is the fourth paragraph.


(And since Akkoma has a bug that uses br br instead of p, and I'm not defending Akkoma on that, but that made the poem-splitting behavior more visible.)


Now, let's consider the same original roses-poem as if it were input from the gem side.
The intent of my commit here is to turn that into


    <p>This is one paragraph.</p>
    <p>This is a second paragraph.</p>

    <p>Roses are red,</br>
    violets are blue,</br>
    the third paragraph's a poem</br>
    through and through.</p>

    <p>This is the fourth paragraph.</p>

Comment (snan) [reply front/text/plain/convert.go:136]: Maybe you're saying that


    This is one paragraph.

    This is a second paragraph.

    Roses are red,
    violets are blue,
    the third paragraph's a poem
    through and through.

    This is the fourth paragraph.

should turn into:

    This is one paragraph.

    This is a second paragraph.

    Roses are red, violets are blue, the third paragraph's a poem through and through.

    This is the fourth paragraph.

I think it's reasonable to argue that **FromHTML** should do that. Maybe. I think it shouldn't, but a case could be made. But I don't think **ToHTML** should do that. Gemini users are used to their newlines being hard.

Comment (dimkr) [reply front/text/plain/convert.go:136]: > Gemini users are used to their newlines being hard.

Yes, and Mastodon users expect the opposite, but doing what looks more like gemtext was a mistake and your PR can fix this. I'll write tests on top of this branch, based on the FromHTML() tests - if this PR's ToHTML() can "reverse" what it does, it's perfect.

Comment (snan) [reply front/text/plain/convert.go:136]: Nice! I modded some of the tests already.

File: front/text/plain/convert_test.go [hunk 2]
```diff
@@ -145,7 +145,7 @@ func TestToHTML_Plain(t *testing.T) {
 
 func TestToHTML_LineBreak(t *testing.T) {
 	post := "this is a line\nthis is another line"
-	expected := `<p>this is a line</p><p>this is another line</p>`
+	expected := `<p>this is a line<br/>this is another line</p>`
```
Comment (dimkr) [inline front/text/plain/convert_test.go:148]: Bug:

`"\n\n\nthis is a line\nthis is another line"`

becomes

`"<p></p><p><br/>this is a line<br/>this is another line</p>"`

Comment (snan) [reply front/text/plain/convert_test.go:148]: What would you want it to be?

Comment (snan) [reply front/text/plain/convert_test.go:148]: <p></br></br></br>this is a line</br>this is another line<p>

If you think that'll look neater, I'll make it so 💁🏻‍♀️

File: front/text/plain/convert.go [hunk 3]
```diff
@@ -132,11 +132,12 @@ func ToHTML(text string) string {
 		return text
 	}
```
Comment (dimkr) [inline front/text/plain/convert.go:134]: The strings.Split() call can be replaced with something more efficient if `lines` is unused

File: front/text/plain/convert_test.go (hunk 2 above)
Comment (dimkr) [reply front/text/plain/convert_test.go:148]: `"<br/><br/><br/><p>this is a line<br/>this is another line</p>"` makes the most sense, no?

Comment (snan) [reply front/text/plain/convert_test.go:148]: > `"<br/><br/><br/><p>this is a line<br/>this is another line</p>"` makes the most sense, no?

Unfortunately not since <br/> isn't a block level element. It needs to be "inside" block level elements like p or h1.

Comment (dimkr) [reply front/text/plain/convert_test.go:148]: Oh no

Comment (snan) [reply front/text/plain/convert_test.go:148]: Pretty happy with the semantics of the most recent commit. It also wraps the whole thing in <p> no matter how many lines there are, again to fit in with the block level aspect of the HTML spec. And that cuts out the string splitting that you wanted to cut, too 👍🏻

Comment (dimkr) [reply front/text/plain/convert_test.go:148]: Small thing: move the MustCompile() call to an init() function, so it's called only once and panics before main() if it fails

**Comment by snan** [comment]
I had missed to para wrap the tests. Also I didn't see this in time before I sent that.

> lol, wait, I want to write tests that check for the desired output first, let them fail, then adjust the code until they pass - every time I change one line in your PR one test passes but another starts failing

Writing tests before having grokked consistent semantics can lead to issues with tests contradicting each other.

The PR at 88a39ed preserves newlines from gem to HTML with a mix of paras and breaks.

```




foo




bar

baz
qux



```

becomes:

    <p><br/><br/><br/><br/>foo<br/><br/><br/></p><p>bar</p><p>baz<br/>qux<br/><br/><br/><br/></p>

The two last newlines in an "inner" sequence of newlines become a paragraph separator sequence, all other newlines become br, and then wrap the whole thing in a pair of paragraph tags.

I thought that was what you wanted: preserve newlines.

The PR at a7b8ecb collapses newlines but doesn't join paras.

So that becomes:

    <p>foo</p><p>bar</p><p>baz<br/>qux</p>

**Comment by dimkr** [comment]
All tests pass except those with leading line breaks.

The output is not faithful HTML representation of the input (I changed the expected output so the tests pass):

```
func TestToHTML_LeadingLineBreak(t *testing.T) {
	post := "\nthis is a line\nthis is another line"
	expected := `<p>this is a line<br/>this is another line</p>`

	html := ToHTML(post)
	assert.Equal(t, expected, html)
}

func TestToHTML_LeadingLineBreaks(t *testing.T) {
	post := "\n\n\nthis is a line\nthis is another line"
	expected := `<p>this is a line<br/>this is another line</p>`

	html := ToHTML(post)
	assert.Equal(t, expected, html)
}
```

**Comment by snan** [comment]
With the a7b8ecb commit I changed it to collapse white space because I thought you thought it was ugly otherwise 🤷🏻‍♀️ 

If you want to preserve the leading line breaks and such then revert a7b8ecb. I'll do that if you ask—I'm learning my lesson to not second guess things now so I'm not messing with it otherwise 🤷🏻‍♀️

**Comment by dimkr** [comment]
I think I'm happy with snan/tootik#1, the way it is now. Text delimited by multiple line breaks becomes a series of paragraphs, single line breaks become `<br/>` and both leading and trailing line breaks are preserved. The result HTML is visually consistent with the plain input, minus the collapsed line breaks between paragraphs (that's what Mastodon does).

**Comment by dimkr** [comment]
Squashed together with snan/tootik#1 into 4d07c98
